pkg/common/validation/overrides.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2022 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validation
+
+import (
+	"fmt"
+
+	operatorv1 "github.com/tigera/operator/api/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/tigera/operator/pkg/common/k8svalidation"
+	"github.com/tigera/operator/pkg/components"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// ValidateContainer is a function that validates the given container.
+type ValidateContainer func(container corev1.Container) error
+
+// NoContainersDefined is a container validation function that is used when no container is expected.
+var NoContainersDefined ValidateContainer = func(container corev1.Container) error {
+	return fmt.Errorf("container %q is invalid. No containers are expected", container.Name)
+}
+
+// ValidateReplicatedPodResourceOverrides validates the given replicated pod resource overrides.
+// validateContainerFn and validateInitContainerFn are used to validate the container overrides.
+func ValidateReplicatedPodResourceOverrides(overrides components.ReplicatedPodResourceOverrides, validateContainerFn ValidateContainer, validateInitContainerFn ValidateContainer) error {
```

Comment:
Maybe a nice cleanup for a future PR would be to instead define an interface type for the validation funcs here:

```go

type OverrideValidator interface {
  ValidateContainer func() err
  ValidateInitContainer func() err
}
```

It makes it fairly easy to extend this function in the future as overrides grows. This is very minor and would recommend against doing that in this PR, but maybe someday if we need to add a third argument to this function an interface would be a clean way to do it.